students/Gregg/lesson5/mailroom.py

```diff
@@ -0,0 +1,268 @@
+
+    The columns are name, total, number and average
+    """
+    row_list = []
+    headers = 'Donor name', 'Total Given', 'Num Gifts', 'Average Gift'
+    format_column_header = "{:<26}|{:^13}|{:^13}|{:^13}".format
+    row_list.append(format_column_header(*headers))
+    row_list.append('_' * 68)
+
+    donor_rows_list = [get_row(donor) for donor in donors]
+    donor_rows_list.sort(key=key1, reverse=True)
+
+    format_row = "{:<27}${:>12.2f} {:>12}  ${:>12.2f}".format
+    row_list.extend([
+        format_row(*row)
+        for row in donor_rows_list
+    ])
+    report_string = ('\n').join(row_list)
+    return report_string
+
+
+def print_report():
+    """Print report to prompt"""
+    print(create_donation_report(donations))
+
+
+def key1(iterable):
+    """Sort key for creating report"""
+    return iterable[1]
+
+
+def get_row(donor):
+    """retrieve the statistics for a given donor"""
+    donors_donations = donations_from(donor)
+    total = sum(donors_donations)
+    num = len(donors_donations)
+    try:
+        avg = total / num
+    except ZeroDivisionError as E:
+        avg = 0
+    return donor, total, num, avg
+
+
+def donations_from(donor):
+    """Return a list of all amounts made by donor ordered chronologically"""
+    # Including this in anticipation that the data struct might change
+    return donations[donor]
+
+
+def send_letters():
+    """Send letters to the donors in the DB that have made donations"""
+    for donor in donations:
+        send_letter(donor)
+
+
+def send_letter(donor):
+    """Send a thank you letter to a donor for the most recent donation"""
+    last_don = last_donation(donor)
+    now = datetime.datetime.now().strftime("%m.%d.%Y.%H.%M.%S")
+    letter_filename = "{}_{}.txt".format(donor, now)
+    try:
+        donor_thx_string = thankyou_string(donor, last_don)
+    except TypeError as E:
+        print((
+            "{1} is in our database but has not made a donation.\n"
+            "No thank you was saved for {1}"
+        ).format(donor)
+        )
+    else:
+        with open(letter_filename, 'w') as letter:
+            letter.write(donor_thx_string)
+        print(f'Sent letter to {donor}')
+
+
+def last_donation(donor):
+    """Return the value of the last donation made by donor"""
+    try:
+        last_donation = donations_from(donor)[-1]
+    except IndexError as E:
+        last_donation = None
+    # This is currently pointelss because the default dict took care of the
+    # the situation where a donor would need to be added seperatley from their
+    # donation. I will leave it in anyway in case it comes again in the future
+    return last_donation
+
+
+primary_switch_dict = {
+    'TY': prompt_thank_you,
+    'CR': print_report,
+    'SL': send_letters
+}
+
+
+def tests():
+    """Test the functions"""
+    pass
+
+
+if __name__ == "__main__":
+    tests()
+    print("Welcome to Gregg's Mailroom 1.3\n")
+    primary_menu()
```

Comment:
Looks good - nice use of exceptions and getting a comprehension in there.  Same suggestions for Lesson 4 mailroom apply here.  Let me know if you have any questions.